crates/types/src/storage.rs

```diff
@@ -8,178 +8,180 @@
 // the Business Source License, use of this software will be governed
 // by the Apache License, Version 2.0.
 
+use crate::epoch::EpochMetadata;
+use crate::errors::GenericError;
 use crate::logs::metadata::Logs;
 use crate::nodes_config::NodesConfiguration;
 use crate::partition_table::FixedPartitionTable;
-use crate::serde::{SerializationError, VersionedSerializer};
 use bytes::BufMut;
+use std::mem;
+
+#[derive(Debug, thiserror::Error)]
+pub enum StorageSerdeError {
+    #[error("failed reading codec: {0}")]
+    ReadingCodec(String),
+    #[error("encoding failed: {0}")]
+    EncodeValue(GenericError),
+    #[error("decoding failed: {0}")]
+    DecodeValue(GenericError),
+    #[error("unsupported codec: {0}")]
+    UnsupportedCodec(StorageCodec),
+}
 
-#[derive(strum_macros::FromRepr)]
-#[repr(u16)]
-pub enum NodesConfigurationVersion {
-    FlexbufferSerde,
+#[derive(Debug, strum_macros::FromRepr, derive_more::Display)]
+#[repr(u8)]
+pub enum StorageCodec {
+    // plain old protobuf
+    Protobuf,
+    // flexbuffers + serde
+    FlexbuffersSerde,
 }
 
-impl From<NodesConfigurationVersion> for u16 {
-    fn from(value: NodesConfigurationVersion) -> Self {
-        value as u16
+impl From<StorageCodec> for u8 {
+    fn from(value: StorageCodec) -> Self {
+        value as u8
     }
 }
 
-impl TryFrom<u16> for NodesConfigurationVersion {
-    type Error = SerializationError;
+impl TryFrom<u8> for StorageCodec {
+    type Error = StorageSerdeError;
 
-    fn try_from(value: u16) -> Result<Self, Self::Error> {
-        NodesConfigurationVersion::from_repr(value).ok_or(SerializationError::ReadingVersion(
-            format!("unknown discriminant '{}'", value),
-        ))
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        StorageCodec::from_repr(value).ok_or(StorageSerdeError::ReadingCodec(format!(
+            "unknown discriminant '{}'",
+            value
+        )))
     }
 }
 
-pub struct NodesConfigurationSerializer;
-
-impl VersionedSerializer for NodesConfigurationSerializer {
-    type Value = NodesConfiguration;
-    type Version = NodesConfigurationVersion;
+pub trait StorageSerde {
+    fn codec() -> StorageCodec;
 
-    fn current_version() -> Self::Version {
-        NodesConfigurationVersion::FlexbufferSerde
+    fn encode<B: BufMut>(&self, buf: &mut B) -> Result<(), StorageSerdeError> {
+        // write codec
+        buf.put_u8(Self::codec().into());
+        // write value
+        self.encode_value(buf)
```

Comment:
Potentially risky design choice, a shared behavior allows implementors to override it and we almost certainly don't want that to be a possibility. An alternative is to split the envelope encoder into an empty struct with the encode/decode functions implemented, and only let StorageSerde trait be responsible for value encoding/decoding (or splitting it into two)
Example:
```rust
enum StorageCodecKind { ... }
struct StorageCodec {}
impl StorageCodec {
   pub fn encode<T: StorageEncode, B: BufMut>(value: &T, buf: &mut B) -> Result<(), ...>
   pub fn decode<T: StorageDecode, B: AsRef[u8]>(buf: B) -> Result<T, ...>
}
```
And types need to implement
```rust
pub trait StorageEncode {
 const DEFAULT_CODEC: StorageCodecKind;
  fn encode_value<B: BufMut>(value: &Self, buf: &mut B) -> Result<(), ...>
}

pub trait StorageDecode {
   fn decode(codec_kind: StorageCodecKind, buf: &[u8]) -> Result<Self, ...>
}
```

You can of course put encode/decode in a single trait instead of splitting them.

Reply:
I think it makes sense to split it as you've suggested. I will update the code accordingly.